File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m=len(grid)
        n=len(grid[0])
        discovered=[[False]*(n) for _ in range(m)]
        res=0
        def dfs(i,j):
            # Out of bounds
            if i<0 or j<0 or i>=m or j>=n:
                return 
            if discovered[i][j]:
                return 
            if grid[i][j]=="0":
                return
            discovered[i][j]=True
            dfs(i+1,j)
            dfs(i-1,j)
            dfs(i,j-1)
            dfs(i,j+1)
        for r in range(m):
            for c in range(n):
                if not discovered[r][c] and grid[r][c]=="1":
                    dfs(r,c)
                    res+=1
        return res


        

File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    bool dfs(vector<vector<char>>& G, int r, int c) {
        int R = G.size();
        int C = G[0].size();
        if (r < 0 or c < 0 or r >= R or c >= C)
            return false;
        if (G[r][c] == '0')
            return false;
        G[r][c] = '0';
        dfs(G, r + 1, c);
        dfs(G, r - 1, c);
        dfs(G, r, c + 1);
        dfs(G, r, c - 1);
        return true;
    }
    int numIslands(vector<vector<char>>& G) {
        ios_base::sync_with_stdio(false);
        cin.tie(nullptr);
        int R = G.size();
        int C = G[0].size();
        int res = 0;
        for (int r = 0; r < R; r++)
            for (int c = 0; c < C; c++) {
                if (G[r][c] == '1') {
                    dfs(G, r, c);
                    res++;
                }
            }
        return res;
    }
};